rholang/src/main/scala/coop/rchain/rholang/interpreter/PrettyPrinter.scala

```diff
@@ -34,112 +36,123 @@ case class PrettyPrinter(
-        (buildString(p1) + " %% " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" %% ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EPlusBody(EPlus(p1, p2)) =>
-        (buildString(p1) + " + " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" + ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EPlusPlusBody(EPlusPlus(p1, p2)) =>
-        (buildString(p1) + " ++ " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" ++ ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EMinusBody(EMinus(p1, p2)) =>
-        (buildString(p1) + " - " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" - ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EMinusMinusBody(EMinusMinus(p1, p2)) =>
-        (buildString(p1) + " -- " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" -- ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EAndBody(EAnd(p1, p2)) =>
-        (buildString(p1) + " && " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" && ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EOrBody(EOr(p1, p2)) =>
-        (buildString(p1) + " || " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" || ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EEqBody(EEq(p1, p2)) =>
-        (buildString(p1) + " == " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" == ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case ENeqBody(ENeq(p1, p2)) =>
-        (buildString(p1) + " != " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" != ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EGtBody(EGt(p1, p2)) =>
-        (buildString(p1) + " > " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" > ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EGteBody(EGte(p1, p2)) =>
-        (buildString(p1) + " >= " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" >= ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case ELtBody(ELt(p1, p2)) =>
-        (buildString(p1) + " < " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" < ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case ELteBody(ELte(p1, p2)) =>
-        (buildString(p1) + " <= " + buildString(p2)).wrapWithBraces
+        (buildStringM(p1) |+| pure(" <= ") |+| buildStringM(p2)).map(_.wrapWithBraces)
       case EMatchesBody(EMatches(target, pattern)) =>
-        (buildString(target) + " matches " + buildString(pattern)).wrapWithBraces
+        (buildStringM(target) |+| pure(" matches ") |+| buildStringM(pattern))
+          .map(_.wrapWithBraces)
       case EListBody(EList(s, _, _, remainder)) =>
-        "[" + buildSeq(s) + buildRemainderString(remainder) + "]"
+        pure("[") |+| buildSeq(s) |+| buildRemainderString(remainder) |+| pure("]")
```

Comment:
this is another pattern that would work for list, tuple, set
 
```scala 
def iterableToString(start : String, s : Seq[String], end : String) = pure(start) |+| buildSeq(s) |+| pure(end)
```

...and it might even work for maps with a bit of effort.